junit-platform-engine/src/main/java/org/junit/platform/engine/UniqueId.java

```diff
@@ -82,10 +84,14 @@ public static UniqueId root(String segmentType, String value) {
 
 	private final UniqueIdFormat uniqueIdFormat;
 	private final List<Segment> segments;
+	private transient int hashCode;
+	private transient String toString;
 
 	private UniqueId(UniqueIdFormat uniqueIdFormat, Segment segment) {
 		this.uniqueIdFormat = uniqueIdFormat;
 		this.segments = singletonList(segment);
+		this.hashCode = this.segments.hashCode();
+		this.toString = this.uniqueIdFormat.format(this);
```

Comment:
I'm not sure if we should do this eagerly. Doing so might slow down other use cases. WDYT?

Reply:
This is `String::format`, ain't it?

If yes, this will be blazing fast ... 🔜 https://openjdk.java.net/jeps/348

Reply:
Well, maybe, but it will still be slower than doing nothing... 😉

Reply:
True.

Reply:
> I'm not sure if we should do this eagerly. Doing so might slow down other use cases. WDYT?

Reply:
@sbrannen here's a screenshot from a [flight recording](https://github.com/marschall/junit5-Segment-hashCode/raw/master/src/main/jfr/junit-5.5.2.jfr) from this test https://github.com/marschall/junit5-Segment-hashCode/blob/master/src/test/java/com/github/marschall/junit5/segmenthashcode/ReproducerTest.java

![Screenshot](https://github.com/marschall/junit5-Segment-hashCode/raw/master/src/main/javadoc/Segment-hashCode.png)

Reply:
@sormuras I don't see how this helps. JEP 348 talks about `String#format`, we're calling `UniqueIdFormat#format` which does not end up calling `String#format`. JEP 348 explicitly is only for JDK code. JEP 348 gave up on intrinsifing `Objects#hashCode`.

Reply:
> [...] which does not end up calling `String#format`.

Wasn't sure about that and didn't verified my guess two days ago. Thanks for the heads-up.

> JEP 348 gave up on intrinsifing `Objects#hashCode`.

Hm, that and the information from above point towards computing and storing those values already in the constructor. Iff those values are always used at least once in the lifetime of an instance, methinks.

Reply:
> @sbrannen here's a screenshot from a flight recording

OK. You convinced me.

Reply:
@marcphilipp I believe both of them ultimately always end up being called

`#hashCode`

The hash based collections in:

- org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution
- org.junit.platform.engine.support.descriptor.AbstractTestDescriptor#addChild
- org.junit.vintage.engine.execution.TestRun

`#toString`

- org.junit.platform.launcher.TestIdentifier#from(TestDescriptor)

I have to admit doing it eagerly is a bit of a cop out because there are many options and trade-offs for computing it lazily if it has to be thread safe (I assume it has to be).
- do we want to use `AtomicReference`?
- do we want to use `volatile` and an `AtomicIntegerFieldUpdater`?
- do we want to use `volatile` and a lock, if so
  - do we want a `synchronized` or a `java.util.concurrent.locks.Lock`?
  - how many locks do we want? a single static one? one per object? one per object and field?
- do we want a benign race like `String#hash`?

Reply:
> Iff those values are always used at least once in the lifetime of an instance, methinks.

Not necessarily since some `TestDescriptors` might get pruned from the test plan. Thus, I think we should go for lazy initialization instead. We could reuse the `MemoizingSupplier` that's already in the codebase:
https://github.com/junit-team/junit5/blob/5e41ebe612fd67e905f96f1dd3184a071b65be17/junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/ExtensionValuesStore.java#L164-L192.

Reply:
On the other hand that would require a lot of additional objects... 🤔

Reply:
I think a benign race would be sufficient because if it ends up being used `hashCode()` and `toString()` will be called from a single thread before the potentially multi-threaded execution begins.

Reply:
> do we want a benign race like String#hash?